### PR TITLE
clean dapo tutorial; standardize hardware params

### PR DESCRIPTION
**Pull request by andrewhinh**
Cleans up the DAPO tutorial and standardizes `TrainConfig` hardware params.

---

<a href="https://modal.devinenterprise.com/review/modal-projects/training-gym/pull/417" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: tutorials/tutorial_generator/rl/000_rl_basics.py [hunk 1]
```diff
@@ -297,39 +326,28 @@ def _train_intro():
...
+            actor_num_gpus_per_node=8,
             tensor_model_parallel_size=1,
+            sequence_parallel=False,       
+            rollout_num_gpus=8,
             rollout_num_gpus_per_engine=1,
-            sequence_parallel=False,
             colocate=True,
             num_rollout=10,
             rollout_batch_size=16,
+            n_samples_per_prompt=2,
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/tutorial_generator/rl/000_rl_basics.py:339]: will fix in future PR where we use only model-specific recipes

File: tutorials/tutorial_generator/rl/000_rl_basics.py [hunk 2]
```diff
@@ -128,12 +131,12 @@ def _score_fn_intro():
     granularity such that it's immediately obvious what the failure mode is (if any). Below, we implement
     the following function:
 
-    - Return `-10` if the model was so incompetent that failed to return three lines.
+    - Return -10 if the model was so incompetent that failed to return three lines.
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/000_rl_basics.py]: I think this typo was here before but
```suggestion
    - Return -10 if the model was so incompetent that it failed to return three lines.
```

File: tutorials/tutorial_generator/rl/000_rl_basics.py [hunk 3]
```diff
@@ -220,11 +223,21 @@ class HaikuDataset(HuggingFaceDataset):
...
+            from datasets import load_dataset
+
+            ds = load_dataset(self.hf_repo, self.hf_config, split=self.hf_split)
+            start = min(self.row_offset, len(ds))
+            stop = len(ds) if not self.n_rows else min(start + self.n_rows, len(ds))
+            return ds.select(range(start, stop))
 
     train_dataset = HaikuDataset(n_rows=10)
-    eval_dataset = HaikuDataset(n_rows=5)
+    eval_dataset = HaikuDataset(n_rows=5, row_offset=10)
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/000_rl_basics.py]: I wonder if we could just get away with [Hugging Face's native split functionality](https://huggingface.co/docs/datasets/v1.0.1/splits.html#examples) instead?

```python
train_dataset = HaikuDataset(hf_split="train[:10]")
eval_dataset = HaikuDataset(hf_split="train[10:15]")
```

That way we don't have to define a separate `load` function (and then need to adapt it for dataset API changes down the line)

File: tutorials/tutorial_generator/rl/000_rl_basics.py [hunk 4]
```diff
@@ -281,6 +294,22 @@ def _score_one(example):
     print(f"average score: {base_mean:.1f}")
 
 
+@markdown
+def _rm_fn_intro():
+    """
+    ## Creating a reward function
+
+    To make our scoring function a reward function, we just need to ensure the output produced
+    by the model during training can be parsed and scored. Simple enough.
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/000_rl_basics.py]: Would prefer something a little more descriptive about what `haiku_rm` and `base_model.parse_response` do, maybe:

> To make our scoring function a reward function, we just need to extract the text from the model's response and pass it to our existing `score_haiku`. Simple enough.

File: tutorials/tutorial_generator/rl/003_on_policy_distillation.py [hunk 5]
```diff
@@ -248,13 +187,23 @@ def _dataset_intro():
...
+            from datasets import load_dataset
+
+            ds = load_dataset(self.hf_repo, self.hf_config, split=self.hf_split)
+            start = min(self.row_offset, len(ds))
+            stop = len(ds) if not self.n_rows else min(start + self.n_rows, len(ds))
+            return ds.select(range(start, stop))
 
     train_dataset = MathDataset(n_rows=100)
-    eval_dataset = MathDataset(n_rows=20)
+    eval_dataset = MathDataset(n_rows=20, row_offset=100)
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/003_on_policy_distillation.py]: Same thing here, perhaps:

```python
train_dataset = MathDataset(hf_split="train[:100]")
eval_dataset = MathDataset(hf_split="train[100:120]")
```

File: tutorials/tutorial_generator/rl/005_dapo.py [hunk 6]
```diff
@@ -1,16 +1,15 @@
-# pyright: reportUndefinedVariable=false, reportMissingImports=false
 """Tutorial source for `005_dapo` — parsed by generate_tutorial.py."""
 
 TUTORIAL_METADATA = {
     "framework": "`slime`",
     "cluster_shape": "1 × 8×H100",
-    "summary": "DAPO on math",
+    "summary": "Large-scale RL for everyone",
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/005_dapo.py:6]: This title seems vague to me, ideally I'd want it to mention DAPO and its benefit (longer chains of thought) somewhere

File: tutorials/tutorial_generator/rl/005_dapo.py [hunk 7]
```diff
@@ -162,122 +120,118 @@ def _normalize_answer(answer: str) -> str:
...
+            start = min(self.row_offset, len(ds))
+            stop = len(ds) if not self.n_rows else min(start + self.n_rows, len(ds))
+            return ds.select(range(start, stop))
 
-        with ThreadPoolExecutor(max_workers=max_concurrency) as executor:
-            rows = list(executor.map(_score_one, eval_dataset.load()))
-        mean = sum(r["score"] for r in rows) / len(rows) if rows else float("nan")
-        return mean, rows
+    train_dataset = MathDataset(n_rows=2000)
+    eval_dataset = MathDataset(n_rows=100, row_offset=2000)
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/005_dapo.py]:
```python
train_dataset = MathDataset(hf_split="train[:2000]")
eval_dataset = MathDataset(hf_split="train[2000:2100]")
```

File: tutorials/tutorial_generator/rl/005_dapo.py [hunk 8]
```diff
@@ -162,122 +120,118 @@ def _normalize_answer(answer: str) -> str:
...
+
+
+@notebook_only
+@code
+def _dataset_peek():
+    df = eval_dataset.to_pandas()
+    print(len(df))
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/005_dapo.py]:
```suggestion
    print(f"{len(df)} rows")
```

just to make this count more clear in the cell output

File: tutorials/tutorial_generator/rl/005_dapo.py [hunk 9]
```diff
@@ -296,26 +250,12 @@ async def dapo_overlong_rm(args, sample, **kwargs) -> float:
...
-    - **Overlong buffer of 2048 tokens** (`L_cache=2048`)
-    - **8192-token response cap** (`L_max=8192`)
-
-    This tutorial runs a short 15-rollout job to demonstrate the DAPO training setup.
-    For a more meaningful accuracy gain, increase the rollout count.
+    ## Commence training
+
+    In addition to model-specific parameters, the recipe below also includes
+    [DAPO-specific](https://arxiv.org/abs/2503.14476)
+    modifications. Note that some parameters have been limited for
```
Comment (anli5005) [inline tutorials/tutorial_generator/rl/005_dapo.py]: I would prefer if we still explain what these DAPO-specific modifications are, or at least which parameters enable them; otherwise it's not very clear to me how I'd take this tutorial and adapt it to my own DAPO runs

**Review by anli5005 — APPROVED** [review]
